File: Clients/macOS/ValeniusApp/Sources/ValeniusApp/IpcClient.swift
// Synchronous IPC client — mirrors Clients/Linux/tray/ipc_client.py. Called from the main
// (UI) thread for quick calls, and off-thread for the 5 s poll. Connects to the daemon's
// Unix socket and speaks the shared length-prefixed JSON framing.

import Darwin
import Foundation
import ValeniusShared

let ipcSocketPath = "/var/run/valenius.sock"

enum IpcError: Error, CustomStringConvertible {
    case daemonNotRunning
    case timedOut
    /// isLanConflict: true when this is a pre-connect local-LAN-conflict refusal — the app
    /// shows a blocking alert for this instead of the usual inline error text.
    case failure(String, isLanConflict: Bool)
    case other(String)

    var description: String {
        switch self {
        case .daemonNotRunning: return "The Valenius background service isn't running."
        case .timedOut: return "The background service didn't respond in time."
        case .failure(let m, _): return m
        case .other(let m): return m
        }
    }
}

enum IpcClient {
    static let timeout: TimeInterval = 10

    // MARK: Commands

    static func status() throws -> TunnelStatus {
        try decode(send(PipeCommand(command: .status)))
    }

    static func syncStatus() throws -> TunnelStatus {
        try decode(send(PipeCommand(command: .syncStatus), timeout: 12))
    }

    static func connect(profile: String?) throws {
        try expectOk(send(PipeCommand(command: .connect, profileName: profile)))
    }

    static func disconnect(profile: String?) throws {
        try expectOk(send(PipeCommand(command: .disconnect, profileName: profile)))
    }

    static func uploadConfig(profile: String, content: String) throws {
        try expectOk(send(PipeCommand(command: .uploadConfig, configContent: content, profileName: profile)))
    }

    @discardableResult
    static func getConfigInfo() throws -> ConfigInfo {
        try decode(send(PipeCommand(command: .getConfigInfo)))
    }

    static func setAutoConnect(_ enabled: Bool) throws {
        try expectOk(send(PipeCommand(command: .setAutoConnect, autoConnectEnabled: enabled)))
    }

    static func deleteProfile(_ profile: String) throws {
        try expectOk(send(PipeCommand(command: .deleteProfile, profileName: profile)))
    }

    static func register() throws -> RegistrationResult {
        try decode(send(PipeCommand(command: .register)))
    }

    static func mfaEnrollConfirm(code: String) throws {
        try expectOk(send(PipeCommand(command: .mfaEnrollConfirm, mfaCode: code)))
    }

    static func sendLogs() throws {
        try expectOk(send(PipeCommand(command: .sendLogs)))
    }

    /// Best-effort: tells the daemon the app is closing so the backend shows offline promptly.
    static func notifyOffline() {
        _ = try? send(PipeCommand(command: .notifyOffline), timeout: 2)
    }

    // MARK: Transport

    private static func send(_ cmd: PipeCommand, timeout: TimeInterval = timeout) throws -> PipeResponse {
        let fd = socket(AF_UNIX, SOCK_STREAM, 0)
        guard fd >= 0 else { throw IpcError.other("socket() failed") }
        defer { close(fd) }

        var tv = timeval(tv_sec: Int(timeout), tv_usec: 0)
        setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))
        setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        let pathBytes = ipcSocketPath.utf8CString
        withUnsafeMutableBytes(of: &addr.sun_path) { raw in
            let buf = raw.bindMemory(to: CChar.self)
            for (i, b) in pathBytes.enumerated() where i < buf.count { buf[i] = b }
        }
        let connected = withUnsafePointer(to: &addr) { ptr -> Int32 in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                Darwin.connect(fd, $0, socklen_t(MemoryLayout<sockaddr_un>.size))
            }
        }
        guard connected == 0 else {
            throw (errno == ENOENT || errno == ECONNREFUSED) ? IpcError.daemonNotRunning : IpcError.other("connect() failed (errno \(errno))")
        }

        do {
            let payload = try JSONEncoder().encode(cmd)
            try IpcProtocol.writeMessage(fd: fd, data: payload)
            let respData = try IpcProtocol.readMessage(fd: fd)
            return try JSONDecoder().decode(PipeResponse.self, from: respData)
        } catch let e as IpcProtocolError {
            switch e {
            case .connectionClosed: throw IpcError.daemonNotRunning
            case .ioError(let n) where n == EAGAIN || n == EWOULDBLOCK: throw IpcError.timedOut
            default: throw IpcError.other("\(e)")
            }
        }
    }

    private static func expectOk(_ resp: PipeResponse) throws {
        if !resp.success {
            throw IpcError.failure(resp.error ?? "The operation failed.", isLanConflict: resp.isLanConflict)
        }
    }

    private static func decode<T: Decodable>(_ resp: PipeResponse) throws -> T {
        guard resp.success else {
            throw IpcError.failure(resp.error ?? "The operation failed.", isLanConflict: resp.isLanConflict)
        }
        guard let json = resp.dataJson, let data = json.data(using: .utf8) else {
            throw IpcError.other("Empty response from the background service.")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: Clients/Linux/tray/ipc_client.py
"""Synchronous IPC client for the tray app.

Called from the GLib main thread (GTK) to communicate with the daemon.
"""
from __future__ import annotations

import json
import socket
from typing import Optional

from shared.messages import (
    CommandType, ConfigInfo, PipeCommand, PipeResponse,
    RegistrationResult, TunnelStatus, VersionCheckResult,
)
from shared.protocol import read_message_sync, write_message_sync

SOCKET_PATH = '/run/valenius.sock'
TIMEOUT = 10


class IpcError(Exception):
    pass


class LanConflictError(IpcError):
    """Raised when a Connect is refused because this machine's own local LAN
    overlaps a network reachable through the VPN (or the VPN-assigned IP itself).
    Mirrors Windows' PipeResponse.IsLanConflict -> LanConflictForm split."""
    pass


def _send(cmd: PipeCommand, timeout: int = TIMEOUT) -> PipeResponse:
    try:
        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
            sock.settimeout(timeout)
            sock.connect(SOCKET_PATH)
            write_message_sync(sock, cmd.to_json())
            raw = read_message_sync(sock)
            return PipeResponse.from_json(raw)
    except FileNotFoundError:
        raise IpcError("Daemon is not running (socket not found).")
    except ConnectionRefusedError:
        raise IpcError("Daemon is not running.")
    except socket.timeout:
        raise IpcError("Daemon did not respond in time.")
    except Exception as e:
        raise IpcError(str(e))


def get_status() -> TunnelStatus:
    resp = _send(PipeCommand(Command=CommandType.Status))
    if not resp.Success:
        raise IpcError(resp.Error or "Status failed.")
    return TunnelStatus.from_json(resp.DataJson or '{}')


def connect(profile_name: Optional[str] = None) -> None:
    resp = _send(PipeCommand(Command=CommandType.Connect, ProfileName=profile_name))
    if not resp.Success:
        if resp.IsLanConflict:
            raise LanConflictError(resp.Error or "Connect failed.")
        raise IpcError(resp.Error or "Connect failed.")


def disconnect(profile_name: Optional[str] = None) -> None:
    resp = _send(PipeCommand(Command=CommandType.Disconnect, ProfileName=profile_name))
    if not resp.Success:
        raise IpcError(resp.Error or "Disconnect failed.")


def mfa_enroll_confirm(code: str) -> bool:
    """Returns True if the backend accepted the TOTP code."""
    resp = _send(PipeCommand(Command=CommandType.MfaEnrollConfirm, MfaCode=code))
    if not resp.Success:
        raise IpcError(resp.Error or "MFA confirmation failed.")
    return True


def upload_config(profile_name: str, content: str) -> None:
    resp = _send(PipeCommand(
        Command=CommandType.UploadConfig,
        ProfileName=profile_name,
        ConfigContent=content,
    ))
    if not resp.Success:
        raise IpcError(resp.Error or "Upload failed.")


def get_config_info() -> ConfigInfo:
    resp = _send(PipeCommand(Command=CommandType.GetConfigInfo))
    if not resp.Success:
        raise IpcError(resp.Error or "GetConfigInfo failed.")
    return ConfigInfo.from_json(resp.DataJson or '{}')


def sync_status() -> TunnelStatus:
    """Force an immediate backend heartbeat + update check and return the fresh
    status (the daemon caps this at 8 s, under our 10 s socket timeout)."""
    resp = _send(PipeCommand(Command=CommandType.SyncStatus))
    if not resp.Success:
        raise IpcError(resp.Error or "Sync failed.")
    return TunnelStatus.from_json(resp.DataJson or '{}')


def set_auto_connect(enabled: bool) -> None:
    resp = _send(PipeCommand(Command=CommandType.SetAutoConnect, AutoConnectEnabled=enabled))
    if not resp.Success:
        raise IpcError(resp.Error or "SetAutoConnect failed.")


def delete_profile(profile_name: str) -> None:
    resp = _send(PipeCommand(Command=CommandType.DeleteProfile, ProfileName=profile_name))
    if not resp.Success:
        raise IpcError(resp.Error or "Delete failed.")


def notify_offline() -> None:
    """Best-effort: tells the daemon the tray is closing so it can immediately
    inform the backend (admin panel shows offline without waiting ~7 minutes)."""
    try:
        _send(PipeCommand(Command=CommandType.NotifyOffline), timeout=2)
    except IpcError:
        pass


def send_logs() -> None:
    """Ask the daemon to collect its (Valenius-only, redacted) logs and upload them."""
    resp = _send(PipeCommand(Command=CommandType.SendLogs))
    if not resp.Success:
        raise IpcError(resp.Error or "Failed to send logs.")


def register() -> RegistrationResult:
    resp = _send(PipeCommand(Command=CommandType.Register))
    if not resp.Success:
        raise IpcError(resp.Error or "Register failed.")
    return RegistrationResult.from_json(resp.DataJson or '{}')


def set_backend_url(dns: str) -> Optional[str]:
    """First-run: send the backend server DNS (the daemon prepends https:// and persists it).
    Returns a non-fatal advisory (e.g. server unreachable) on success, or None if all is well.
    Raises IpcError only when the daemon rejects the address (invalid) or is unreachable."""
    resp = _send(PipeCommand(Command=CommandType.SetBackendUrl, BackendDns=dns), timeout=15)
    if not resp.Success:
        raise IpcError(resp.Error or "Could not save the server address.")
    return resp.Error
